File: examples/julia/json_parser_test.jl
# Test runner for json_parser.jl using Julia's built-in `Test` standard
# library. A failing `@testset` exits non-zero, which fails the CI matrix leg.

using Test

include("json_parser.jl")

const SAMPLE = """
{
  "os": ["ubuntu-latest", "windows-latest", "macos-latest"],
  "versions": { "julia": ["1.10", "1.11"] },
  "ghpages_branch": "ghgapes"
}
"""

@testset "json_parser" begin
    config = parse_config(SAMPLE)
    @test config !== nothing

    os = config["os"]
    @test length(os) == 3
    @test "ubuntu-latest" in os
    @test "windows-latest" in os
    @test "macos-latest" in os

    # Assert structure, not specific version values, so bumping the matrix
    # versions never requires editing this test.
    versions = config["versions"]["julia"]
    @test !isempty(versions)
    @test all(v -> !isempty(v), versions)

    @test config["ghpages_branch"] == "ghgapes"

    @test parse_config("not json"; silent = true) === nothing
end

File: examples/dart/json_parser_test.dart
// Dependency-free test runner for json_parser.dart. Uses plain assertions that
// throw on failure (a non-zero exit fails the CI matrix leg), so it needs no
// external test package.

import 'json_parser.dart';

int _failures = 0;

void check(bool condition, String message) {
  if (!condition) {
    _failures++;
    print('FAIL: $message');
  }
}

const sample = '''
{
  "os": ["ubuntu-latest", "windows-latest", "macos-latest"],
  "versions": { "dart": ["3.11.6", "3.12.1"] },
  "ghpages_branch": "ghgapes"
}
''';

void main() {
  final config = parseConfig(sample);
  check(config != null, 'config should not be null');

  final os = (config!['os'] as List).cast<String>();
  check(os.length == 3, 'os should have 3 entries');
  check(os.contains('ubuntu-latest'), 'os should contain ubuntu-latest');
  check(os.contains('windows-latest'), 'os should contain windows-latest');
  check(os.contains('macos-latest'), 'os should contain macos-latest');

  // Assert structure, not specific version values, so bumping the matrix
  // versions never requires editing this test.
  final versions = ((config['versions'] as Map)['dart'] as List).cast<String>();
  check(versions.isNotEmpty, 'dart versions should be a non-empty list');
  check(
    versions.every((v) => v.isNotEmpty),
    'each dart version should be a non-empty string',
  );

  check(config['ghpages_branch'] == 'ghgapes', 'ghpages_branch should match');

  check(parseConfig('not json', silent: true) == null, 'invalid JSON -> null');

  if (_failures > 0) {
    throw StateError('$_failures test(s) failed');
  }
  print('All tests passed');
}
